os/sys/stack-check.c

```diff
@@ -0,0 +1,158 @@
+#include "contiki.h"
+#include "sys/stack-check.h"
+#include "dev/watchdog.h"
+#include <string.h>
+
+#include "sys/log.h"
+#define LOG_MODULE "Stack"
+#define LOG_LEVEL LOG_LEVEL_MAIN
+
+/*---------------------------------------------------------------------------*/
+/* linker will provide a symbol for the end of the .bss segment */
+extern uint8_t _stack;
+/* the approximate starting point of stack for platforms that don't define STACK_ORIGIN */
+static void *stack_origin;
+
+#if STACK_CHECK_PERIODIC_CHECKS
+PROCESS(stack_check_process, "Stack check");
+#endif
+/*---------------------------------------------------------------------------*/
+/* The symbol with which the stack memory is initially filled */
+#define STACK_FILL 0xcd
+/*---------------------------------------------------------------------------*/
+#ifdef STACK_ORIGIN
+/* use the #defined value */
+#define GET_STACK_ORIGIN() STACK_ORIGIN
+#else
+/* use the approximate value */
+#define GET_STACK_ORIGIN() stack_origin
+#endif
+/*---------------------------------------------------------------------------*/
+void
+stack_check_init(void)
+{
+  uint8_t *p;
+
+  stack_origin = &p;
+  p = &_stack;
+
+  /* Note: this is expected to be called before the WDT is started! */
+  while(p < (uint8_t *)stack_origin) {
+    *p++ = STACK_FILL;
+  }
+
+#if STACK_CHECK_PERIODIC_CHECKS
+  /* Start the periodic checker process */
+  process_start(&stack_check_process, NULL);
+#endif
+}
+/*---------------------------------------------------------------------------*/
+uint16_t
+stack_check_get_usage(void)
+{
+  uint8_t *p = &_stack;
+
+  /* Make sure WDT is not triggered */
+  watchdog_periodic();
+
+  /* Skip the bytes used after heap; it's 1 byte by default for _stack,
+   * more than that means dynamic memory allocation is used somewhere.
+   */
+  while(*p != STACK_FILL && p < (uint8_t *)GET_STACK_ORIGIN()) {
+    p++;
+  }
+
+  /* Skip the region of the memory reserved for the stack not used yet by the program */
+  while(*p == STACK_FILL && p < (uint8_t *)GET_STACK_ORIGIN()) {
+    p++;
+  }
+
+  /* Make sure WDT is not triggered */
+  watchdog_periodic();
+
+  if(p >= (uint8_t*)GET_STACK_ORIGIN()) {
+    /* This means the stack is screwed. */
+    return 0xffff;
+  }
+
+  return (uint8_t *)GET_STACK_ORIGIN() - p;
+}
+/*---------------------------------------------------------------------------*/
+uint16_t
+stack_check_get_reserved_size(void)
+{
+  return (uint8_t *)GET_STACK_ORIGIN() - &_stack;
+}
+/*---------------------------------------------------------------------------*/
+#if STACK_CHECK_PERIODIC_CHECKS
+/*---------------------------------------------------------------------------*/
+PROCESS_THREAD(stack_check_process, ev, data)
+{
+  static struct etimer et;
+
+  PROCESS_BEGIN();
+
+  etimer_set(&et, STACK_CHECK_PERIOD);
+
+  while(1) {
+    PROCESS_WAIT_EVENT_UNTIL(etimer_expired(&et));
+
+    uint16_t actual = stack_check_get_usage();
+    uint16_t allowed = stack_check_get_reserved_size();
```

Comment:
code style: I was always assuming we don't accept variables declared in the middle of blocks, but I'm not sure now. Opinions?

Reply:
I'll change for consistency.